lib/protocol/helpers.js

```diff
@@ -0,0 +1,89 @@
+var util =  require('../util');
+var AWS = require('../core');
+
+/**
+ * Prepend prefix defined by API model to endpoint that's already
+ * constructed. This feature does not apply to operations using
+ * endpoint discovery and can be disabled.
+ * @api private
+ */
+function populateHostPrefix(request)  {
+  var enabled = request.service.config.hostPrefixEnabled;
+  if (!enabled) return request;
+  var operationModel = request.service.api.operations[request.operation];
+  //don't marshal host prefix when operation has endpoint discovery traits
+  if (hasEndpointDiscover(request)) return request;
+  if (operationModel.endpoint && operationModel.endpoint.hostPrefix) {
+    var hostPrefixNotation = operationModel.endpoint.hostPrefix;
+    var hostPrefix = expandHostPrefix(hostPrefixNotation, request.params, operationModel.input);
+    prependEndpointPrefix(request.httpRequest.endpoint, hostPrefix);
+    validateHostname(request.httpRequest.endpoint.hostname);
+  }
+  return request
+}
+
+/**
+ * @api private
+ */
+function hasEndpointDiscover(request) {
+  var api = request.service.api;
+  var operationModel = api.operations[request.operation];
+  var isEndpointOperation = api.endpointOperation && (api.endpointOperation === util.string.lowerFirst(operationModel.name));
+  return (operationModel.endpointDiscoveryRequired !== 'NULL' || isEndpointOperation === true);
+}
+
+/**
+ * @api private
+ */
+function expandHostPrefix(hostPrefixNotation, params, shape) {
+  util.each(shape.members, function(name, member) {
+    if (member.hostLabel === true) {
+      if (typeof params[name] !== 'string' || params[name] === '') {
+        throw util.error(new Error(), {
+          message: 'Parameter ' + name + ' should be a non-empty string.',
+          code: 'InvalidParameter'
+        });
+      }
+      var regex = new RegExp('\\{' + name + '\\}');
+      hostPrefixNotation = hostPrefixNotation.replace(regex, params[name]);
```

Comment:
this will only replace the first instance of the hostLabel. While odd, a hostPrefix template could have multiple instances of the same hostLabel.

If you add the global (g) flag to this regex i think it will solve the issue.